### PR TITLE
Add data_dependent_output tag; generalize proxy tensor to test it

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83330
* #83316
* __->__ #83312

Fixes https://github.com/pytorch/pytorch/issues/83251

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: aten/src/ATen/native/tags.yaml [hunk 1]
```diff
@@ -12,6 +12,12 @@
   desc: |
           This tag indicates if an operator's output's shape depends on input Tensor
           data.
+- tag: data_dependent_output
+  desc: |
+          Operator has a non-Tensor output whose value is dependent on the data
```
Comment (albanD) [inline aten/src/ATen/native/tags.yaml:17]: Why the non-Tensor output constraint here?
This is excluding ops like `nonzero()` on purpose?

Comment (ezyang) [reply aten/src/ATen/native/tags.yaml:17]: Yes, nonzero is covered by `dynamic_output_shape`, which with symbolic shapes *is* traceable (but non-Tensor outputs are always not traceable, unless we make them return symbolic bools sometimes).

Comment (albanD) [reply aten/src/ATen/native/tags.yaml:17]: Ok!

**Review by albanD — APPROVED** [review]

File: aten/src/ATen/native/tags.yaml [hunk 2]
```diff
@@ -12,6 +12,12 @@
   desc: |
           This tag indicates if an operator's output's shape depends on input Tensor
           data.
+- tag: data_dependent_output
```
Comment (zou3519) [inline aten/src/ATen/native/tags.yaml:15]: To clarify, does this include operators that may error out depending on input values? (would be good to clarify in the description)

Comment (ezyang) [reply aten/src/ATen/native/tags.yaml:15]: Well, based on what the PR does, the answer is no. But... do you think it should?

**Pull request closed without merging**